Changed to prevent certain DoS attacks (#280)
rsksmart/RSKIPs: IPs/RSKIP243.md

# **Abstract**

This RSKIP proposes eliminating inter-block gas refunds and replace them with a 100% intra-transaction gas refunds. It also proposes to reimburse the user for storage and contracts that are created but cleared in the same transaction avoiding requiring state trie updates and storage operations, increasing platform fairness.

# **Motivation**

> The motivation for the removal of storage refunds can be found in [EIP3298](https://eips.ethereum.org/EIPS/eip-3298). Essentially, gas refunds that span over blocks bring numerous problems. However, instead of removing storage gas reimbursements completely, we propose keeping them but only if the storage creation and removal occurs within a single transaction. By doing so we reduce the gas costof common Solidity language constructions, such as reentrancy semaphores. This change improves the platform fairness because only performs refunds when the state trie is not impacted, and the I/O cost is avoided. 

Reviewer: ```suggestion
The motivation for the removal of storage refunds can be found in [EIP3298](https://eips.ethereum.org/EIPS/eip-3298). Essentially, gas refunds that span over blocks bring numerous problems. However, instead of removing storage gas reimbursements completely, we propose keeping them but only if the storage creation and removal occurs within a single transaction. By doing so we reduce the gas cost of common Solidity language constructions, such as reentrancy semaphores. This change improves the platform fairness because only performs refunds when the state trie is not impacted, and the I/O cost is avoided. 
```

> Currently, when a contract is destroyed, the reimbursed amount is constant. However, the variable cost related to the number of bytes installed can be very high. We propose to reimburse 99% of this amount, also increasing platform fairness.

Reviewer: ```suggestion
Currently, when a contract is destroyed, the reimbursed amount is constant. However, the variable cost related to the number of bytes installed can be very high. We propose to reimburse 99% of this amount, which also increases the platform fairness.
```

Because intra-transaction refunds cannot increase block processing time, the cap to 50% of the consumed gas is removed.

# **Specification**

Starting from an activation block (TBD) storage cleanup refunds are modified both for `SSTORE` and `SELFDESTRUCT`, and the cost of `SSTORE` is also simplified. Refunds can only be collected in the same transaction the gas is spent. 

A map of sets *storage_cells_touched\[contractAddress\]\[storageAddress\]* is created in each call frame.

> The maps is  empty when call begins.

Reviewer: ```suggestion
The map is empty when each call begins.
```

If a call reverts/OOG, the map is discarded.

Each time `SSTORE` is called, its cell address is looked up in *storage_cells_touched* in the current call frame. If it exists, then the cost of `SSTORE` is 400. If it doesn't exists, then it is 20000. 

When a call returns without reverting changes, the contents of *storage_cells_touched* is merged into the caller's map.

At the end of transaction processing, all storage cells that exists in the top-level *storage_cells_touched* are scanned.
For each cell, if its value is equal to their pre-execution value, then 19600 gas is refunded.

> Also at the end of transaction processing, if a contract was selfdestructed and the contract did not exists pre-execution, then 24K gas (out of the 32K originally paid) plus 99% the refund of the variable cost required to install the code (198 per byte) is refunded.

Reviewer: ```suggestion
Also at the end of transaction processing, if a contract was selfdestructed and the contract did not exist prior to execution, then 24K gas (out of the 32K originally paid), plus 99% of the variable cost required to install the code (198 per byte), is refunded.
```

> The cap of the refund to 50% of the transaction gas consumed is removed. 

Reviewer: ```suggestion
The cap of the refund to 50% of the transaction gas consumed is removed.
```
